server/src/main/java/org/elasticsearch/index/mapper/FallbackSyntheticSourceBlockLoader.java

```diff
@@ -80,26 +106,9 @@ public void read(int docId, StoredFields storedFields, Builder builder) throws I
             }
 
             Map<String, List<IgnoredSourceFieldMapper.NameValue>> valuesForFieldAndParents = new HashMap<>();
-
-            // Contains name of the field and all its parents
-            Set<String> fieldNames = new HashSet<>() {
-                {
-                    add("_doc");
-                }
-            };
-
-            var current = new StringBuilder();
-            for (String part : fieldName.split("\\.")) {
-                if (current.isEmpty() == false) {
-                    current.append('.');
-                }
-                current.append(part);
-                fieldNames.add(current.toString());
-            }
-
             for (Object value : ignoredSource) {
-                IgnoredSourceFieldMapper.NameValue nameValue = IgnoredSourceFieldMapper.decode(value);
-                if (fieldNames.contains(nameValue.name())) {
+                IgnoredSourceFieldMapper.NameValue nameValue = (IgnoredSourceFieldMapper.NameValue) value;
```

Comment:
If we fallback to the default `StoredFieldLoader` instead of our custom `IgnoredSourceFieldLoader`, this would cause a `ClassCastException` (except that now our visitor is filtering for `_ignored_source.<field_name>` instead of `_ignored_source`, so the `ignoredSource` variable is null and this whole block is skipped).